### ISSUE TITLE
Is there a trainable critic model in PPO training?

### ISSUE THREAD

**Issue by skepsun** [huggingface/trl#769]
For multiple adapter PPO process, the reward adapter is loaded with `requires_grad=False`. In standard PPO process, the critic model is initialized with the loaded reward model but can still be updated/trained. Does trl combine actor and critic models into one model (with a learnable value head for critic model?) which is updated with an extra `vf_loss`?

Maybe we can create the third adapter `critic_adapter` initialized from the `reward_adapter` but with `requires_grad=True`?

**Comment by lvwerra** [comment]
This is currently not supported. If there is evidence that this helps training we could consider adding it. Do you have any experiments on this?

**Comment by zhang2514yuchi** [comment]
I have the same question because the standard PPO uses GAE to calculate advantages. This step requires using a value, but in your code, you have used the GAE formula without a critic model. May I ask if you are treating the actor's output as the value? Is there a theoretical basis for this?

**Comment by lvwerra** [comment]
I don't know if there is a theoretical basis for that, but it is quite common to use the same base model with an action and value head on top of the base LLM. In principle you can use two separate models which will also double the memory consumption.

**Comment by luo-li-ba-suo** [comment]
Critic model may need to be warmed up before PPO. But it will cause unexpected training on the actor model when the critic and actor model share one base model. We are eager for independent support for the actor and the critic model.

**Comment by Ricardokevins** [comment]
> Critic model may need to be warmed up before PPO. But it will cause unexpected training on the actor model when the critic and actor model share one base model. We are eager for independent support for the actor and the critic model.

I quite agree with you

**Comment by lvwerra** [comment]
We welcome community contributions so if you are up to opening a PR implementing this I'd be happy to review!

**Comment by vwxyzjn** [comment]
Let's move the discussion to https://github.com/huggingface/trl/issues/1063#issuecomment-1875967021. Closing this issue now.
